azurelinuxagent/ga/collect_telemetry_events.py

```diff
@@ -0,0 +1,585 @@
+        """
+        Parse the Json event from file. Raise InvalidExtensionEventError if the event breaches pre-set contract.
+        :param extension_event: The json event from file
+        :return: Verified Json event that qualifies the contract.
+        """
+
+        clean_string = lambda x: x.strip() if x is not None else x
+
+        event_size = 0
+        key_err_msg = "{0}: {1} not found"
+
+        # Convert the dict to all lower keys to avoid schema confusion.
+        # Only pick the params that we care about and skip the rest.
+        event = dict((k.lower(), clean_string(v)) for k, v in extension_event.items() if
+                     k.lower() in self._EXTENSION_EVENT_REQUIRED_FIELDS)
+
+        # Trim message and only pick the first 3k chars
+        message_key = ExtensionEventSchema.Message.lower()
+        if message_key in event:
+            event[message_key] = event[message_key][:self._EXTENSION_EVENT_MAX_MSG_LEN]
+        else:
+            raise InvalidExtensionEventError(
+                key_err_msg.format(InvalidExtensionEventError.MissingKeyError, ExtensionEventSchema.Message))
+
+        if not event[message_key]:
+            raise InvalidExtensionEventError(
+                "{0}: {1} should not be empty".format(InvalidExtensionEventError.EmptyMessageError,
+                                                     ExtensionEventSchema.Message))
+
+        for required_key in self._EXTENSION_EVENT_REQUIRED_FIELDS:
+            # If all required keys not in event then raise
+            if required_key not in event:
+                raise InvalidExtensionEventError(
+                    key_err_msg.format(InvalidExtensionEventError.MissingKeyError, required_key))
+
+            # If the event_size > _EXTENSION_EVENT_MAX_SIZE=6k, then raise
+            if event_size > self._EXTENSION_EVENT_MAX_SIZE:
+                raise InvalidExtensionEventError(
+                    "{0}: max event size allowed: {1}".format(InvalidExtensionEventError.OversizeEventError,
+                                                              self._EXTENSION_EVENT_MAX_SIZE))
+
+            event_size += len(event[required_key])
+
+        return event
+
+    @staticmethod
+    def _replace_or_add_param_in_event(event, replace_or_add_params):
+        for param in event.parameters:
+            if param.name in replace_or_add_params:
+                param.value = replace_or_add_params.pop(param.name)
+
+        if not replace_or_add_params:
+            # All values replaced, return
+            return
+
+        # Add the remaining params to the event
+        for param_name in replace_or_add_params:
+            event.parameters.append(TelemetryEventParam(param_name, replace_or_add_params[param_name]))
+
+
+class _CollectAndEnqueueEventsPeriodicOperation(PeriodicOperation):
+    """
+    Periodic operation to collect and send telemetry events located in the events folder
```

Comment:
Suggestion to clarify what "send" means -> to the common event queue, not to wireserver.

Reply:
Modified the comments